### PR TITLE
Feature/hover participants

### PR DESCRIPTION
**Pull request by MidnightJava**
Refactor hover participants. Reference Brainwy tracker #653.

Passed all Junit and JUnit Plug-in tests, except for IronPython (not nested) and known/expected test failures.

### CODE REVIEW

**Comment by fabioz** [comment]
I still haven't taken a look at this one, but I wanted to ask first if the other pull requests are deprecated now or now (if they are, please close them, otherwise, just let me know and I'll take a look at those too).

**Comment by MidnightJava** [comment]
Yes the other ones are deprecated by this one. I thought I closed #155, but
found it still open. I closed #155 and #157 just now. The former is no
longer needed and the commit for the latter (add lib-dynload to PYTHONPATH
for tests) was incorporated into PR #159.

-Mark

File: plugins/org.python.pydev/src/org/python/pydev/utils/Messages.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * Licensed under the terms of the Eclipse Public License (EPL).
+ * Please see the license.txt included with this distribution for details.
+ * Any modifications to this file must keep this entire header intact.
+ * 
+ * Copied from the JDT implementation of
+ * <code>org.eclipse.jdt.internal.corext.util.Messages</code>.
+ */
+package org.python.pydev.utils;
+
+import com.ibm.icu.text.MessageFormat;
```
Comment (fabioz) [inline plugins/org.python.pydev/src/org/python/pydev/utils/Messages.java]: I think that we don't need the com.ibm.icu dependency: you should be able to use the standard String.format() for this (code from JDT may be before java 1.5, but we're on java 7 now anyways, so the com.ibm.icu dependency should probably be removed).

File: plugins/org.python.pydev/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -41,7 +41,10 @@ Require-Bundle: org.eclipse.ui,
  org.python.pydev.shared_ui;bundle-version="[4.5.2,4.5.3)",
  org.python.pydev.shared_core;bundle-version="[4.5.2,4.5.3)",
  org.eclipse.e4.ui.css.swt.theme;resolution:=optional,
- org.eclipse.e4.ui.services;resolution:=optional
+ org.eclipse.e4.ui.services;resolution:=optional,
+ com.ibm.icu,
```
Comment (fabioz) [inline plugins/org.python.pydev/META-INF/MANIFEST.MF]: Remove com.ibm.icu dependency.

File: plugins/org.python.pydev/src/org/python/pydev/editor/hover/PyEditorHoverConfigurationBlock.java [hunk 3]
```diff
@@ -0,0 +1,1018 @@
+/**
+ * Copyright (c) 2016 by Brainwy Software LTDA. All Rights Reserved.
```
Comment (fabioz) [inline plugins/org.python.pydev/src/org/python/pydev/editor/hover/PyEditorHoverConfigurationBlock.java]: If it was copied, you should keep the original copyright (the important thing is that it's EPL).

File: plugins/org.python.pydev/src/org/python/pydev/utils/PyEditorMessages.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.ResourceBundle;
+
+import org.eclipse.osgi.util.NLS;
+
+/**
+ * Helper class to get NLSed messages.
+ */
+public final class PyEditorMessages extends NLS {
+
+    private static final String BUNDLE_FOR_CONSTRUCTED_KEYS = "org.eclipse.jdt.internal.ui.javaeditor.ConstructedJavaEditorMessages";//$NON-NLS-1$
```
Comment (fabioz) [inline plugins/org.python.pydev/src/org/python/pydev/utils/PyEditorMessages.java]: The BUNDLE_FOR_CONSTRUCTED_KEYS is still referencing org.eclipse.jdt.

**Comment by fabioz** [comment]
I've gotten the pull request here, but after applying it, the hover for the debug and the markers are not appearing anymore...

I.e.: 

Given a program with:

```
def check(aa):
    print(aa) # <- add breakpoint here (hovering over aa should show int: 1 + the definition location)
    undefined_var_err_here # Hovering here should show the error message

check(1)
```

Adding the breakpoint in the given line and running in debug, when hovering over aa or over the undefined_var_err_here, I'm not getting the expected results.

**Comment by MidnightJava** [comment]
I have the same issue with the debug hover, but the marker hover works for me. Can you please chack that it's the highest priority hover (i.e. smallest number) enabled in non-combining mode? In combining mode, make sure it has no higher priority hover with preempt set to true. When I have an error in the code, I see the same info when hovering over the invalid code as I do when hovering over the error marker in the margin.

For the debug hover, I don't think I understand the runtime lifecycle. I'm setting the hovers from the new extension point in PyEditConfiguration#gettextHover(ISourceViewer, String, int). This method is invoked for each state mask whenever I launch an editor, and I see the debug hover set as the hover if so configured. There is no additional call to this method when I hover over a debug breakpoint, but the debug hover that was selected previously when this method was called is not invoked. Can you tell me where PyDev looks for a text hover in debug mode when hovering over a breakpoint?

**Comment by MidnightJava** [comment]
I think I found the problem with the debug hover. Ther ewere two problems.

1. Since PyDebugHover#getHoverInfo is not called from the UI thread, I had to wrap a call within it (when getting the text viewer selection) in Display#syncExec
2. I had a bug in PyDebugHover#isContentTypeSupported, and it was returning the wrong value.

Please let me know if the debug hover is now working correctly for you, and also about the marker hover.

**Comment by fabioz** [comment]
Can you redo this pull request targeting the master branch? (I deleted the "development" branch as all the code should be done targeting the "master" branch now -- sorry for the trouble, unfortunately github has no way to change it, although there's quite an old bug requesting it: https://github.com/isaacs/github/issues/18)

**Pull request closed without merging**